src/experimental/ProbabilisticGraphicalModels/bayesnet.jl

```diff
@@ -184,11 +199,98 @@ If Z is provided, the conditioning information in `bn` will be ignored.
 function is_conditionally_independent end
 
 function is_conditionally_independent(bn::BayesianNetwork{V}, X::V, Y::V) where {V}
-    # TODO: Implement
+    # Use currently observed variables as Z
+    Z = V[v for (v, is_obs) in zip(bn.names, bn.is_observed) if is_obs]
+    return is_conditionally_independent(bn, X, Y, Z)
 end
 
 function is_conditionally_independent(
     bn::BayesianNetwork{V}, X::V, Y::V, Z::Vector{V}
 ) where {V}
-    # TODO: Implement
+    # Get vertex IDs
+    x_id = bn.names_to_ids[X]
+    y_id = bn.names_to_ids[Y]
+    z_ids = Set([bn.names_to_ids[z] for z in Z])
+
+    # Track visited nodes and their states
+    n_vertices = nv(bn.graph)
+    visited = falses(n_vertices)
+
+    # Queue entries are (node_id, from_parent)
+    queue = Tuple{Int,Bool}[]
+
+    # Start from X
+    push!(queue, (x_id, true))   # As if coming from parent
+    push!(queue, (x_id, false))  # As if coming from child
+
+    while !isempty(queue)
+        current_id, from_parent = popfirst!(queue)
+
+        if visited[current_id]
+            continue
+        end
+        visited[current_id] = true
+
+        # If we reached Y, path is active
+        if current_id == y_id
+            return false
+        end
+
+        is_conditioned = current_id in z_ids
+        parents = inneighbors(bn.graph, current_id)
+        children = outneighbors(bn.graph, current_id)
+
+        # Case 1: Node is not conditioned
+        if !is_conditioned
+            # Can go to children if coming from parent or at start node
+            if from_parent || current_id == x_id
+                for child in children
+                    push!(queue, (child, true))
+                end
+            end
+
+            # Can go to parents if coming from child or at start node
+            if !from_parent || current_id == x_id
+                for parent in parents
+                    push!(queue, (parent, false))
+                end
+            end
+        end
+
+        # Case 2: Node is conditioned or has conditioned descendants
+        if is_conditioned || has_conditioned_descendant(bn, current_id, z_ids)
+            # If this is a collider or descendant of collider
+            if length(parents) > 1 || !isempty(children)
+                # Can go to parents regardless of direction
+                for parent in parents
+                    push!(queue, (parent, false))
+                end
+            end
+        end
+    end
+
+    return true
+end
+
+function has_conditioned_descendant(bn::BayesianNetwork, node_id::Int, z_ids::Set{Int})
```

Comment:
slight confusion here, why the ball can pass through if the collider child has conditioned descendants? and this could also incur repeated computations?

Reply:
[https://pmc.ncbi.nlm.nih.gov/articles/PMC6089543/figure/F4/](url)
![image](https://github.com/user-attachments/assets/4bfdf571-d13a-4205-8dcb-98d34ade2317)

I am considering this. Please let me know what do you think.

Reply:
Yes I think it might incur repeated computations, but its a trade off for ensuring all possible paths are considered. I am thinking of keeping a visited nodes list, but unsure if that will cause conflict

Reply:
Ah i see makes sense